### PR TITLE
Adds "domain" to signing logic

### LINKED ISSUE #29: Define the builder signing API in to include a signing domain and fork version etc.
from @protolambda:

Define the signing API in a way where you are forced to include a signing domain and fork version etc. 

Signatures should be constructed in a way where they are unique for the purpose of the work and version of the protocol. See [`compute_signing_root`](https://github.com/ethereum/consensus-specs/blob/v1.1.10/specs/phase0/beacon-chain.md#compute_signing_root) and [`get_domain`](https://github.com/ethereum/consensus-specs/blob/v1.1.10/specs/phase0/beacon-chain.md#get_domain) in the phase0 beacon spec. 

You essentially want to compute a "domain" (32 bytes) from a fork digest (constructed from genesis + fork info) and a type-domain (e.g. proposals and attestations have different domains to be sure they never have the same signature for two different intents). Or you can implement your own non-standard domain method. But definitely don't just create a BLS signature over data that might have a different meaning later on in the protocol. cc @lightclient

_Originally posted by @protolambda in https://github.com/protolambda/mergemock/pull/28#discussion_r871381698_

---

See also https://github.com/ethereum/consensus-specs/pull/2884 for the DomainType

### PR DESCRIPTION
**Pull request by lightclient**
fixes #29

### CODE REVIEW

**Comment by metachris** [comment]
That looks great! 

(ping @terencechain @Ruteri)

File: types/signing.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+type HashTreeRoot interface {
+	HashTreeRoot() ([32]byte, error)
+}
+
+func ComputeDomain(dt DomainType, forkVersion uint32, genesisValidatorsRoot *Root) [32]byte {
+	if genesisValidatorsRoot == nil {
+		var tmp Root
+		genesisValidatorsRoot = &tmp
+	}
+	forkDataRoot, _ := (&forkData{
```
Comment (terencechain) [inline types/signing.go:44]: I've noticed here the error is neglected, but given this is a mock, I am not concerned

**Review by terencechain — APPROVED** [review]
lgtm!

File: types/signing.go (hunk 1 above)
Comment (lightclient) [reply types/signing.go:44]: Right, I briefly looked at the generated code for HTR here and it didn't _seem_ like it was possible to error.

**Pull request merged**